filters/include/pcl/filters/experimental/crop_box.h

```diff
@@ -0,0 +1,192 @@
+   */
+  inline void
+  setMin(const Eigen::Vector4f& min_pt)
+  {
+    min_pt_ = min_pt;
+  }
+
+  /** \brief Get the value of the minimum point of the box, as set by the user
+   * \return the value of the internal \a min_pt parameter.
+   */
+  inline Eigen::Vector4f
+  getMin() const
+  {
+    return min_pt_;
+  }
+
+  /** \brief Set the maximum point of the box
+   * \param[in] max_pt the maximum point of the box
+   */
+  inline void
+  setMax(const Eigen::Vector4f& max_pt)
+  {
+    max_pt_ = max_pt;
+  }
+
+  /** \brief Get the value of the maximum point of the box, as set by the user
+   * \return the value of the internal \a max_pt parameter.
+   */
+  inline Eigen::Vector4f
+  getMax() const
+  {
+    return max_pt_;
+  }
+
+  /** \brief Set a translation value for the box
+   * \param[in] translation the (tx,ty,tz) values that the box should be translated by
+   */
+  inline void
+  setTranslation(const Eigen::Vector3f& translation)
+  {
+    translation_ = translation;
+  }
+
+  /** \brief Get the value of the box translation parameter as set by the user. */
+  Eigen::Vector3f
+  getTranslation() const
+  {
+    return translation_;
+  }
+
+  /** \brief Set a rotation value for the box
+   * \param[in] rotation the (rx,ry,rz) values that the box should be rotated by
+   */
+  inline void
+  setRotation(const Eigen::Vector3f& rotation)
+  {
+    rotation_ = rotation;
+  }
+
+  /** \brief Get the value of the box rotatation parameter, as set by the user. */
+  inline Eigen::Vector3f
+  getRotation() const
+  {
+    return rotation_;
+  }
+
+  /** \brief Set a transformation that should be applied to the cloud before filtering
+   * \param[in] transform an affine transformation that needs to be applied to the
+   * cloud before filtering
+   */
+  inline void
+  setTransform(const Eigen::Affine3f& transform)
+  {
+    transform_ = transform;
+  }
+
+  /** \brief Get the value of the transformation parameter, as set by the user. */
+  inline Eigen::Affine3f
+  getTransform() const
+  {
+    return transform_;
+  }
+
+  /** \brief Sample of point indices
+   * \param[out] indices the resultant point cloud indices
+   */
+  void
+  applyFilter(Indices& indices) override
+  {
+    Eigen::Affine3f box_transform;
+    pcl::getTransformation(translation_(0),
+                           translation_(1),
+                           translation_(2),
+                           rotation_(0),
+                           rotation_(1),
+                           rotation_(2),
+                           box_transform);
+    const Eigen::Matrix4f pt_transform =
+        (box_transform.inverse() * transform_).matrix();
+
+    const auto lambda = [&](const PointCloud& cloud, index_t idx) {
+      const Eigen::Vector4f pt = pt_transform * cloud.at(idx).getVector4fMap();
+      return (pt.array() >= min_pt_.array()).template head<3>().all() &&
+             (pt.array() <= max_pt_.array()).template head<3>().all();
+    };
```

Comment:
Do you think a check for pt_transform being identity would help in increasing the speed a tiny bit?

Reply:
Possibly, but I think that would be rare in real-world scenarios. More ideas: 1. make pt_transform a 3x4 matrix because the last element of pt is not used anyway. 2. if pt_transform is translation only or rotation only, compute only what is necessary instead of full matrix-vector-multiplication (these ideas can be explored later though)

Reply:
> Possibly, but I think that would be rare in real-world scenarios.

I usually transform my point cloud first so multiple downstream logic can run independently on it. So all crop-box/filter operations happen with no transform.

Reply:
> Having no TF might not be so rare

I agree this is not so rare. From my experience working with lidar, I use `CropBox` to just remove the points far from the origin without any TFs, just to reduce the runtime for downstream tasks.

Reply:
> 1. make pt_transform a 3x4 matrix because the last element of pt is not used anyway.

I just benchmark it, and surprisingly found 4x3` is a bit slower. I guess it is due to the following conditions check can be run with SSE

```cpp
return (pt.array() >= min_pt_.array()).template head<3>().all() &&
       (pt.array() <= max_pt_.array()).template head<3>().all();
```

`pt` will be `Vector3f` with 3x4 matrix which **may** not trigger SSE

## Benchmark

### 3x4

```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
320p_CropBox               1.50 ms         1.50 ms          461
320p_FunctorCropBox        1.35 ms         1.35 ms          580
480p_CropBox               3.03 ms         3.03 ms          230
480p_FunctorCropBox        2.72 ms         2.72 ms          253
720p_CropBox               7.93 ms         7.90 ms           67
720p_FunctorCropBox        8.25 ms         8.24 ms          118
1080p_CropBox              20.8 ms         20.8 ms           49
1080p_FunctorCropBox       18.5 ms         18.5 ms           37
1440p_CropBox              38.6 ms         38.5 ms           27
1440p_FunctorCropBox       33.2 ms         33.2 ms           26
```

### 4x4

```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
320p_CropBox               1.57 ms         1.57 ms          445
320p_FunctorCropBox        1.28 ms         1.28 ms          674
480p_CropBox               3.03 ms         3.03 ms          229
480p_FunctorCropBox        2.46 ms         2.46 ms          273
720p_CropBox               7.88 ms         7.88 ms           67
720p_FunctorCropBox        7.55 ms         7.54 ms          134
1080p_CropBox              21.1 ms         21.1 ms           48
1080p_FunctorCropBox       17.0 ms         17.0 ms           40
1440p_CropBox              38.3 ms         38.2 ms           27
1440p_FunctorCropBox       30.6 ms         30.6 ms           28
```